File: L1/Z3/libLOOP.py
def factorial(n):
    result = 1
    while not n==0:
        result *= n
        n -= 1
    return result

def gcd(a,b):

    while not b == 0:
        c = a
        a = b
        b = c%b

    return a
    


def extended_euclidean(a,b):
    if b==0:
        return a,1,0
    
    coeff = []
    i = 0

    while not b == 0:
        coeff.append((a,b))
        c = a
        a = b
        b = c%b 
        i += 1

    i -= 1

    x = 1
    y = 0

    while i >= 0:
        a_new, b_new = coeff[i]
        xold = x
        yold = y
        x = y
        y = xold - (a_new//b_new)*yold
        i -= 1

    return a, x, y



def eqDIO(a,b,c):
    gcdd, x, y = extended_euclidean(a,b)
    if c%gcdd == 0:
        return x*c//gcdd, y*c//gcdd
    else:
        return "no solutions"
    
    

File: L1/Z1/libLOOP.c
#include "lib.h"
#include "stdio.h"


int factorial(int n)
{
    int result = n;
    while(n != 1)
    {
        n -= 1;
        result *= n;
    }
    return result;
}

int gcd(int a, int b)
{
    if(a<b)
    {
        int c = b;
        b = a;
        a = c;
    }

    while(b!=0)
    {
       int c = a%b;
       a = b;
       b = c;
    }

    return a;
}

int extended_euclidean(int a, int b, int *x, int *y)
{
    pair coefficients[400];

    int i = 0;
    while(b!=0)
    {
        pair p;
        p.first = a;
        p.second = b;
        coefficients[i] = p;

        int c = a;
        a = b;
        b = c%b;
        i++;
    }
    
    //adding last pair (gcd,0)
    pair p;
    p.first = a;
    p.second = b;
    coefficients[i] = p;

    //solutions for (gcd * x1 + 0 * y1 = gcd)
    int x1 = 1;
    int y1 = 0;
    i--;

    while(i>=0)
    {
        int x1old = x1;
        int y1old = y1;
        x1 = y1old;
        y1 = x1old - (coefficients[i].first/coefficients[i].second)*y1old;
        i--;
    }

    *x = x1;
    *y = y1;
    return a;
}

pair eqDIOFAN(int a, int b, int c)
{
    pair result;

    int x, y;
    int gcd = extended_euclidean(a, b, &x, &y);

    if(c % gcd != 0)
    {
        result.first = -1;
        result.second = -1;
        result.valid = false;
        return result;
    }
    else 
    {
        result.first = x*c/gcd;
        result.second = y*c/gcd;
        result.valid = true;
        return result;
    }
}